### PR TITLE
Refactor and clean up attn_dp shardings

### PR DESCRIPTION
**Pull request by wenxindongwork**
# Description

Refactor some attention dp shardings. Impact to performance should be positive and minimal.  

Previously inputs are all-gathered before the embedding layer, now this all-gather is eliminated. 

Previously output hidden state is all-gathered during logits computation, now this all-gather happens in the step_fn. 

# Tests

Verified with e2e profiling. 

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/common/sharding.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ class ShardingAxisNameBase:
     MOE_TENSOR = ('attn_dp', 'model')
     EXPERT = ('attn_dp', 'expert', 'model')
     EXPERT_DATA = ('data', 'attn_dp', 'expert', 'model')
-    VOCAB = ('model', 'attn_dp', 'expert')
+    VOCAB = ('attn_dp', 'model', 'expert')
```
Comment (bzgoogle) [inline tpu_inference/layers/common/sharding.py:44]: may i know the reason of switching this 2 axises for VOCAB?

**Review by bzgoogle — APPROVED** [review]

File: tpu_inference/layers/common/sharding.py (hunk 1 above)
Comment (wenxindongwork) [reply tpu_inference/layers/common/sharding.py:44]: it's just for cosmetics :) everywhere else had attn_dp before model
